### PR TITLE
Doc comments, module expose/hide, base bound changes

### PR DESCRIPTION
**Pull request by harendra-kumar**

### CODE REVIEW

File: core/src/Streamly/Internal/Unicode/Array/Char.hs [hunk 1]
```diff
@@ -6,6 +6,11 @@
 -- Maintainer  : streamly@composewell.com
 -- Stability   : experimental
 -- Portability : GHC
+
+-- XXX Remove this or move this to Unicode.Utf32 making "Array Char" as a
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Unicode/Array/Char.hs:10]: `Array Char` can be an alias for `Utf32` perhaps.

File: src/Streamly/Internal/Unicode/Utf8.hs [hunk 2]
```diff
@@ -5,7 +5,12 @@
 -- Maintainer  : streamly@composewell.com
 -- Stability   : experimental
 -- Portability : GHC
---
+
+-- XXX We can move this to stream-core/streamly-unicode-core, and provide an
+-- additional module in streamly-unicode for case conversions (because it
+-- depends on unicode-data). Or just keep all of it in streamly-unicode
+-- which will have a dependency on unicode-data.
```
Comment (adithyaov) [inline src/Streamly/Internal/Unicode/Utf8.hs:12]: We can just do `streamly-unicode` first and further split if required.

File: core/src/Streamly/Internal/Data/Either/Strict.hs [hunk 3]
```diff
@@ -19,6 +19,8 @@
 --
 module Streamly.Internal.Data.Either.Strict
     ( Either' (..)
+    -- XXX Remove these and add lazyEither/strictEither to convert to and from
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Either/Strict.hs:22]: Why remove these? We need to check for perf. Converting to lazy `Either` and checking adds another level of if/else which might be less performant.

Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Either/Strict.hs:22]: We can remove these accordingly.

File: core/streamly-core.cabal [hunk 4]
```diff
@@ -355,6 +355,7 @@ library
                      , Streamly.Data.Fold.Tee
                      , Streamly.Data.Parser
                      , Streamly.Data.Unfold
+                     , Streamly.Data.Array
```
Comment (adithyaov) [inline core/streamly-core.cabal:382]: We are exposing this. We might need to review the API.

File: core/streamly-core.cabal [hunk 5]
```diff
@@ -364,6 +363,8 @@ library
                      , Streamly.FileSystem.Handle
                      , Streamly.Console.Stdio
 
+    other-modules: Streamly.Data.Stream.Zip
```
Comment (adithyaov) [inline core/streamly-core.cabal:391]: We can also hide the `Fold.Tee` type, eventually.

File: core/src/Streamly/Internal/Data/Either/Strict.hs (hunk 3 above)
Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Either/Strict.hs:22]: > Why remove these? We need to check for perf. Converting to lazy `Either` and checking adds another level of if/else which might be less performant.

It should get simplified by the compiler. We can check that.

File: core/streamly-core.cabal (hunk 4 above)
Comment (harendra-kumar) [reply core/streamly-core.cabal:382]: Yes, we will review. I had taken a look and it looked ok.

**Pull request merged**